File: VIM-BE/src/main/java/com/bezkoder/spring/login/sa/bll/servicesimpl/CustomFormApplicationService.java
package com.bezkoder.spring.login.sa.bll.servicesimpl;

import java.util.List;
import org.apache.log4j.LogManager;
import org.apache.log4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import com.bezkoder.spring.login.sa.bll.services.ICustomFormApplicationService;
import com.bezkoder.spring.login.sa.dal.dao.ICfgTblCustomFormApplicationDAO;
import com.bezkoder.spring.login.sa.dal.entities.CfgTblCustomFormApplication;

@Service
public class CustomFormApplicationService implements ICustomFormApplicationService {

    @Autowired
    private ICfgTblCustomFormApplicationDAO customFormApplicationDAO;

    private Logger logger = LogManager.getLogger(CustomFormApplicationService.class);

    public CustomFormApplicationService() {
    }

    @Override
    public List<CfgTblCustomFormApplication> getAllApplications() {
        logger.debug("getAllApplications()");
        return customFormApplicationDAO.getAllApplications();
    }

    @Override
    public List<CfgTblCustomFormApplication> getApplicationsByFormId(Integer formId) {
        logger.debug("getApplicationsByFormId() - formId: " + formId);
        return customFormApplicationDAO.getApplicationsByFormId(formId);
    }

    @Override
    public List<CfgTblCustomFormApplication> getApplicationsByUserId(Integer userId) {
        logger.debug("getApplicationsByUserId() - userId: " + userId);
        return customFormApplicationDAO.getApplicationsByUserId(userId);
    }

    @Override
    public List<CfgTblCustomFormApplication> getDepartmentApplications(Integer userId) {
        logger.debug("getDepartmentApplications() - userId: " + userId);
        return customFormApplicationDAO.getDepartmentApplications(userId);
    }

    @Override
    public CfgTblCustomFormApplication getApplicationById(Integer applicationId) {
        logger.debug("getApplicationById() - applicationId: " + applicationId);
        return customFormApplicationDAO.getApplicationById(applicationId);
    }

    @Override
    public String submitApplication(CfgTblCustomFormApplication application) {
        logger.debug("submitApplication()");
        return customFormApplicationDAO.submitApplication(application);
    }

    @Override
    public String updateApplication(CfgTblCustomFormApplication application) {
        logger.debug("updateApplication()");
        return customFormApplicationDAO.updateApplication(application);
    }

    @Override
    public String updateApplicationPdf(Integer applicationId, byte[] pdfData, String pdfName, String pdfMime,
            boolean refreshCapfSignatures, boolean suppressEditNotification) {
        logger.debug("updateApplicationPdf() - applicationId: " + applicationId + ", refreshCapfSignatures: "
                + refreshCapfSignatures + ", suppressEditNotification: " + suppressEditNotification);
        return customFormApplicationDAO.updateApplicationPdf(applicationId, pdfData, pdfName, pdfMime,
                refreshCapfSignatures, suppressEditNotification);
    }

    @Override
    public String deleteApplication(Integer applicationId) {
        logger.debug("deleteApplication() - applicationId: " + applicationId);
        return customFormApplicationDAO.deleteApplication(applicationId);
    }

    @Override
    public List<CfgTblCustomFormApplication> getApplicationsByStatus(String status) {
        logger.debug("getApplicationsByStatus() - status: " + status);
        return customFormApplicationDAO.getApplicationsByStatus(status);
    }

    @Override
    public List<CfgTblCustomFormApplication> getApplicationsApprovedByUser(String status, Integer userId) {
        logger.debug("getApplicationsApprovedByUser() - status: " + status + ", userId: " + userId);
        return customFormApplicationDAO.getApplicationsApprovedByUser(status, userId);
    }

    @Override
    public List<CfgTblCustomFormApplication> getApplicationsByStatusAndUserId(String status, Integer userId) {
        logger.debug("getApplicationsByStatusAndUserId() - status: " + status + ", userId: " + userId);
        return customFormApplicationDAO.getApplicationsByStatusAndUserId(status, userId);
    }

    @Override
    public String getNextApplicationCode(Integer formId) {
        logger.debug("getNextApplicationCode() - formId: " + formId);
        return customFormApplicationDAO.getNextApplicationCode(formId);
    }

    @Override
    public List<CfgTblCustomFormApplication> getApplicationsPendingApprovalForDepartmentHead(
            Integer departmentHeadUserId) {
        logger.debug(
                "getApplicationsPendingApprovalForDepartmentHead() - departmentHeadUserId: " + departmentHeadUserId);
        return customFormApplicationDAO.getApplicationsPendingApprovalForDepartmentHead(departmentHeadUserId);
    }

    @Override
    public List<CfgTblCustomFormApplication> getAllApplicationsPendingApproval() {
        logger.debug("getAllApplicationsPendingApproval()");
        return customFormApplicationDAO.getAllApplicationsPendingApproval();
    }

    @Override
    public String approveApplication(Integer applicationId, String remarks) {
        logger.debug("approveApplication() - applicationId: " + applicationId);
        return approveApplication(applicationId, remarks, null, "SYSTEM", null);
    }

    @Override
    public String approveApplication(Integer applicationId, String remarks, Integer approverUserId, String approvedVia,
            String approvedIp) {
        logger.debug("approveApplication() - applicationId: " + applicationId + ", approverUserId: " + approverUserId);
        return customFormApplicationDAO.approveApplication(applicationId, remarks, approverUserId, approvedVia,
                approvedIp);
    }

    @Override
    public String rejectApplication(Integer applicationId, String remarks) {
        logger.debug("rejectApplication() - applicationId: " + applicationId);
        return customFormApplicationDAO.rejectApplication(applicationId, remarks);
    }

    @Override
    public String sendBackApplication(Integer applicationId, String remarks) {
        logger.debug("sendBackApplication() - applicationId: " + applicationId);
        return customFormApplicationDAO.sendBackApplication(applicationId, remarks);
    }

    @Override
    public String sendBackApplicationToInitiator(Integer applicationId, String remarks) {
        logger.debug("sendBackApplicationToInitiator() - applicationId: " + applicationId);
        return customFormApplicationDAO.sendBackApplicationToInitiator(applicationId, remarks);
    }

    @Override
    public String resubmitApplicationFromInitiator(Integer applicationId, String remarks, Integer userId) {
        logger.debug("resubmitApplicationFromInitiator() - applicationId: " + applicationId + ", userId: " + userId);
        return customFormApplicationDAO.resubmitApplicationFromInitiator(applicationId, remarks, userId);
    }

    @Override
    public String requestApplicationOpinion(Integer applicationId, Integer opinionUserId, String remarks) {
        logger.debug("requestApplicationOpinion() - applicationId: " + applicationId + ", opinionUserId: " + opinionUserId);
        return customFormApplicationDAO.requestApplicationOpinion(applicationId, opinionUserId, remarks);
    }

    @Override
    public String submitApplicationOpinion(Integer applicationId, String action, String remarks) {
        logger.debug("submitApplicationOpinion() - applicationId: " + applicationId + ", action: " + action);
        return customFormApplicationDAO.submitApplicationOpinion(applicationId, action, remarks);
    }

    @Override
    public String sendSubmissionEmailsForApplication(Integer applicationId) {
        logger.debug("sendSubmissionEmailsForApplication() - applicationId: " + applicationId);
        return customFormApplicationDAO.sendSubmissionEmailsForApplication(applicationId);
    }

    @Override
    public String sendTemplatePostApprovalEmails(Integer applicationId) {
        logger.debug("sendTemplatePostApprovalEmails() - applicationId: " + applicationId);
        return customFormApplicationDAO.sendTemplatePostApprovalEmails(applicationId);
    }

    @Override
    public String sendTemplatePostApprovalEmails(Integer applicationId, byte[] initiatorPdf, byte[] approverPdf,
            String pdfName, String pdfMime) {
        logger.debug("sendTemplatePostApprovalEmails() - applicationId: " + applicationId + ", custom PDFs");
        return customFormApplicationDAO.sendTemplatePostApprovalEmails(applicationId, initiatorPdf, approverPdf,
                pdfName, pdfMime);
    }

    @Override
    public String assignAssetCode(Integer applicationId, String assetCode, Integer userId, String approvedIp) {
        logger.debug("assignAssetCode() - applicationId: " + applicationId + ", userId: " + userId);
        return customFormApplicationDAO.assignAssetCode(applicationId, assetCode, userId, approvedIp);
    }

    @Override
    public String assignPrCode(Integer applicationId, String prCode, Integer userId, String approvedIp) {
        logger.debug("assignPrCode() - applicationId: " + applicationId + ", userId: " + userId);
        return customFormApplicationDAO.assignPrCode(applicationId, prCode, userId, approvedIp);
    }

    @Override
    public String assignPoCode(Integer applicationId, String poCode, Integer userId, String approvedIp) {
        logger.debug("assignPoCode() - applicationId: " + applicationId + ", userId: " + userId);
        return customFormApplicationDAO.assignPoCode(applicationId, poCode, userId, approvedIp);
    }
}

File: VIM-FE/src/app/services/custom-form-application/custom-form-application.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { urls } from 'src/app/utils/urls';

@Injectable({
  providedIn: 'root'
})
export class CustomFormApplicationService {

  constructor(
    private http: HttpClient
  ) { }

  submitApplication(payload: any) {
    return this.http.post(urls.API_URL + 'submitApplication', payload);
  }

  getAllApplications() {
    return this.http.get(urls.API_URL + 'getAllApplications');
  }

  getApplicationsByFormId(formId: number) {
    return this.http.get(urls.API_URL + 'getApplicationsByFormId?formId=' + formId);
  }

  getApplicationsByUserId(userId: number) {
    return this.http.get(urls.API_URL + 'getApplicationsByUserId?userId=' + userId);
  }

  getApplicationById(applicationId: number) {
    return this.http.get(urls.API_URL + 'getApplicationById?applicationId=' + applicationId);
  }

  updateApplication(payload: any) {
    const sanitized = this.stripTransientFields(payload);
    return this.http.post(urls.API_URL + 'updateApplication', sanitized);
  }

  updateApplicationPdf(applicationId: number, pdfBlob: Blob, filename?: string, refreshCapfSignatures?: boolean) {
    const formData = new FormData();
    formData.append('applicationId', String(applicationId));
    formData.append('pdf', pdfBlob, filename || 'application.pdf');
    if (refreshCapfSignatures) {
      formData.append('refreshCapfSignatures', 'true');
    }
    return this.http.post(urls.API_URL + 'updateApplicationPdf', formData);
  }

  deleteApplication(applicationId: number) {
    return this.http.post(urls.API_URL + 'deleteApplication?applicationId=' + applicationId, {});
  }

  getApplicationsByStatus(status: string) {
    return this.http.get(urls.API_URL + 'getApplicationsByStatus?status=' + status);
  }

  getApplicationsByStatusAndUserId(status: string, userId: number) {
    return this.http.get(urls.API_URL + `getApplicationsByStatusAndUserId?status=${status}&userId=${userId}`);
  }

  getApplicationsApprovedByUser(status: string, userId: number) {
    return this.http.get(urls.API_URL + `getApplicationsApprovedByUser?status=${status}&userId=${userId}`);
  }

  getNextApplicationCode(formId: number) {
    return this.http.get(urls.API_URL + 'getNextApplicationCode?formId=' + formId);
  }

  getApplicationsPendingApproval(departmentHeadUserId: number) {
    return this.http.get(urls.API_URL + 'getApplicationsPendingApproval?departmentHeadUserId=' + departmentHeadUserId);
  }

  getAllApplicationsPendingApproval() {
    return this.http.get(urls.API_URL + 'getAllApplicationsPendingApproval');
  }

  approveApplication(applicationId: number, remarks?: string, approverUserId?: number) {
    const body: any = {
      applicationId: applicationId,
      remarks: remarks || ''
    };
    if (approverUserId != null) {
      body.approverUserId = approverUserId;
    }
    return this.http.post(urls.API_URL + 'approveApplication', body);
  }

  rejectApplication(applicationId: number, remarks?: string) {
    return this.http.post(urls.API_URL + 'rejectApplication', {
      applicationId: applicationId,
      remarks: remarks || ''
    });
  }

  sendBackApplication(applicationId: number, remarks: string) {
    return this.http.post(urls.API_URL + 'sendBackApplication', {
      applicationId: applicationId,
      remarks: remarks
    });
  }

  sendBackToInitiator(applicationId: number, remarks: string) {
    return this.http.post(urls.API_URL + 'sendBackToInitiator', {
      applicationId: applicationId,
      remarks: remarks
    });
  }

  sendSubmissionEmails(applicationId: number) {
    return this.http.post(urls.API_URL + 'sendSubmissionEmails?applicationId=' + applicationId, {});
  }

  assignAssetCode(applicationId: number, assetCode: string, userId?: number) {
    return this.http.post(urls.API_URL + 'assignAssetCode', null, {
      params: {
        applicationId: applicationId as any,
        assetCode: assetCode,
        userId: userId ?? ''
      }
    });
  }

  assignPrCode(applicationId: number, prCode: string, userId?: number) {
    return this.http.post(urls.API_URL + 'assignPrCode', null, {
      params: {
        applicationId: applicationId as any,
        prCode: prCode,
        userId: userId ?? ''
      }
    });
  }

  private stripTransientFields(payload: any): any {
    if (!payload || typeof payload !== 'object') return payload;
    const sanitized = { ...payload };
    // Frontend-only flag; backend entity does not allow it.
    if ('isCapfForm' in sanitized) {
      delete (sanitized as any).isCapfForm;
    }
    return sanitized;
  }
}
